src/xrpld/app/misc/detail/LendingHelpers.cpp

```diff
@@ -0,0 +1,1654 @@
+    // paid for a full payment
+    auto const accruedInterest = detail::loanAccruedInterest(
+        rawPrincipalOutstanding,
+        periodicRate,
+        parentCloseTime,
+        startDate,
+        prevPaymentDate,
+        paymentInterval);
+    XRPL_ASSERT(
+        accruedInterest >= 0,
+        "ripple::detail::computeFullPaymentInterest : valid accrued interest");
+
+    auto const prepaymentPenalty =
+        tenthBipsOfValue(rawPrincipalOutstanding, closeInterestRate);
+    XRPL_ASSERT(
+        prepaymentPenalty >= 0,
+        "ripple::detail::computeFullPaymentInterest : valid prepayment "
+        "interest");
+
+    return accruedInterest + prepaymentPenalty;
+}
+
+Number
+calculateFullPaymentInterest(
+    Number const& periodicPayment,
+    Number const& periodicRate,
+    std::uint32_t paymentRemaining,
+    NetClock::time_point parentCloseTime,
+    std::uint32_t paymentInterval,
+    std::uint32_t prevPaymentDate,
+    std::uint32_t startDate,
+    TenthBips32 closeInterestRate)
+{
+    Number const rawPrincipalOutstanding =
+        detail::loanPrincipalFromPeriodicPayment(
+            periodicPayment, periodicRate, paymentRemaining);
+
+    return calculateFullPaymentInterest(
+        rawPrincipalOutstanding,
+        periodicRate,
+        parentCloseTime,
+        paymentInterval,
+        prevPaymentDate,
+        startDate,
+        closeInterestRate);
+}
+
+LoanState
+calculateRawLoanState(
+    Number const& periodicPayment,
+    Number const& periodicRate,
+    std::uint32_t const paymentRemaining,
+    TenthBips32 const managementFeeRate)
+{
+    if (paymentRemaining == 0)
+    {
+        return LoanState{
+            .valueOutstanding = 0,
+            .principalOutstanding = 0,
+            .interestDue = 0,
+            .managementFeeDue = 0};
+    }
+    Number const rawValueOutstanding = periodicPayment * paymentRemaining;
+    Number const rawPrincipalOutstanding =
+        detail::loanPrincipalFromPeriodicPayment(
+            periodicPayment, periodicRate, paymentRemaining);
+    Number const rawInterestOutstanding =
+        rawValueOutstanding - rawPrincipalOutstanding;
+    Number const rawManagementFeeOutstanding =
+        tenthBipsOfValue(rawInterestOutstanding, managementFeeRate);
+
+    return LoanState{
+        .valueOutstanding = rawValueOutstanding,
+        .principalOutstanding = rawPrincipalOutstanding,
+        .interestDue = rawInterestOutstanding - rawManagementFeeOutstanding,
+        .managementFeeDue = rawManagementFeeOutstanding};
+};
+
+LoanState
+calculateRawLoanState(
+    Number const& periodicPayment,
+    TenthBips32 interestRate,
+    std::uint32_t paymentInterval,
+    std::uint32_t const paymentRemaining,
+    TenthBips32 const managementFeeRate)
+{
+    return calculateRawLoanState(
+        periodicPayment,
+        loanPeriodicRate(interestRate, paymentInterval),
+        paymentRemaining,
+        managementFeeRate);
+}
+
+LoanState
```

Comment:
Why is it `Rounded`? There is no any rounding in this function.

Reply:
> Why is it `Rounded`? There is no any rounding in this function.

Are you talking about this function?
```
LoanState
constructRoundedLoanState(
    Number const& totalValueOutstanding,
    Number const& principalOutstanding,
    Number const& managementFeeOutstanding)
```

You're right: There's nothing in the function that checks or enforces rounding, but it's intended for places where the _inputs_ are already rounded, usually because they're coming in from the `Loan` object.

I renamed this overload to just `constructLoanState` in 27a7138f881bc3369b78b39f70a48910f33b888d